### PR TITLE
Refuse merge to non master branch

### LINKED ISSUE #6: Refuse merge if the target branch of a pull request is not the integration branch
If you create a pull request against a different branch than the integration branch (master), merging it would merge the entire thing into master. If you are working on a series of dependent PRs, that could create problems. We should bail out on PRs that are created against a different branch than master. We could add better support for dependent PRs later.

### PR DESCRIPTION
**Pull request by adrianotm**
This PR solves the issue of rejecting commands that are run on PRs that don't have "master" as the base branch. But, if you create a PR with a base branch "a", and then change the base branch to "master", the bot will still reject the commands, as the `PullRequest` data will be out of date. Solving [this issue](https://github.com/channable/hoff/issues/86) would allow updating the base branch when the PR gets edited. Once that issue is implemented, this PR will be ready to merge. Fixes #6

### CODE REVIEW

**Comment by marcin-rzeznicki** [comment]
Thanks but: why are we _assigneess_?  do you want any of us to review this once ready?

**Comment by marcin-rzeznicki** [comment]
BTW I am personally not a fan of commit messages such as: _remove comment_ or _add a test_ as they're meaningless to understand the history of project. Maybe you could rebase it some time later and e.g. squash _fix tests_ with the previous commit etc. We normally use `fixup! xxx` for this as `hoff` autosquashes when merging and the history looks nice. Cheers!

**Comment by adrianotm** [comment]
I made a mistake with the assigneess. :sweat_smile:  I'd like you to review this once it's done.

**Comment by joris-burgers** [comment]
Quick Github tip: If you put `Fixes #6` in the description, the issue that is attached will automatically be closed once this PR is merged. It also provides a quick reference to the issue this PR fixes.

File: src/Logic.hs
Comment (FPtje) [inline src/Logic.hs:378]: The `master` branch is hardcoded here, but it is already part of the config. See `ProjectConfiguration.branch`

File: src/Project.hs
Comment (FPtje) [inline src/Project.hs:87]: I'm not sure `Forbidden` is the right term here. Forbidden can be for any reason. In this case it is because the base branch is not set, but one can also imagine the Forbidden state meaning there are orphan fixup commits (see #37). Would `IncorrectBaseBranch` be a better name?

File: src/Logic.hs
Comment (FPtje) [inline src/Logic.hs:584]: Also hardcoded `master` here. I think it's easiest to just leave it out here.

```suggestion
  PrStatusForbidden  -> "Pull request refused: the target branch must be the integration branch."
```

File: src/Logic.hs
Comment (FPtje) [inline src/Logic.hs:208]: I see that several constructors and functions are getting two `Branch` arguments. One for the to-be-merged branch, and one for the base branch. To prevent mixing the two arguments, would it be an idea to create a `newtype BaseBranch`? 

On one hand a base branch is conceptually also `Branch`, but on the other hand there's also a `RemoteBranch` already 🤔

This is an open question, having two `Branch` parameters might not be _that_ bad. Having a `BaseBranch` would probably mean some conversion function would need to exist to convert it to `Branch`. It's a tradeoff.

File: src/Project.hs
Comment (FPtje) [inline src/Project.hs:78]: Nit: spelling error in `unablet`

**Review by FPtje — COMMENTED** [review]
Nice work! I like the approach. As discussed, here's some initial feedback :+1:

File: src/Logic.hs
Comment (marcin-rzeznicki) [reply src/Logic.hs:208]: `RemoteBranch` is for the thingies that absolutely have to reference something in `origin`. I have not yet reviewed (I thought it was still a draft), so I am not sure if this is one of such cases, but it could be

File: src/Logic.hs
Comment (adrianotm) [reply src/Logic.hs]: I think I've resolved this.

File: src/Logic.hs
Comment (adrianotm) [reply src/Logic.hs:208]: I think `RemoteBranch` is not needed for this case.

Making a new type for the base branch would make things clearer, but I'm not sure whether it's worth to complicate things.

Comment (joris-burgers) [reply src/Logic.hs:208]: > Making a new type for the base branch would make things clearer, but I'm not sure whether it's worth to complicate things.

My 2 cents would be that it would be worth it, either a type-alias or a newtype. I prefer the newtype, because not only does the type document itself, but it prevents you from accidentally swapping the arguments. You can now swap the arguments and the type-system will not catch it. If we add `newtype`, than it would be a type-error to get the arguments mixed-up.

Comment (adrianotm) [reply src/Logic.hs:208]: Good point!

File: src/Logic.hs
Comment (marcin-rzeznicki) [inline src/Logic.hs:330]: Good point!

File: tests/Spec.hs
Comment (marcin-rzeznicki) [inline tests/Spec.hs:712]: Nit: it's not so much the _pull request_ that is being refused here, but _merge_

File: src/Logic.hs
Comment (marcin-rzeznicki) [inline src/Logic.hs:254]: Can we have a test of that wrt to changing the base branch? So basically that something like `[CommentAdded prId "deckard" "@bot merge", PullRequestEdited prId "blah blah" (Branch "master"), CommentAdded prId "deckard" "@bot merge" ]` results in the PR **not** rejected?

**Review by marcin-rzeznicki — COMMENTED** [review]
Good job, I like it! Just one nit and request for some additional testing

File: src/Git.hs
Comment (marcin-rzeznicki) [inline src/Git.hs:109]: Is the other one (`(Sha, Branch)`) still used then?

Comment (adrianotm) [reply src/Git.hs:109]: It is used here: https://github.com/channable/hoff/blob/abbf862e9d9c46c1cce64f4c6fae3fe2e37b112b/src/Git.hs#L286-L296

I think `ForcePush` is used for the testing branch as well, so I can't change the `Branch` to `BaseBranch`.

Comment (marcin-rzeznicki) [reply src/Git.hs:109]: :ok_hand:

File: src/Git.hs
Comment (joris-burgers) [inline src/Git.hs:75]: Nit: Can you use haddock style comments here, so `-- | A type ...`

File: tests/Spec.hs
Comment (joris-burgers) [inline tests/Spec.hs:237]: Just a general comment in this file, the lines are getting quite long, could you maybe introduce some newlines or move some things to variables (like a global `masterBranch = BaseBranch "master"` at the top of this file, that would safe a lot of redundant code).

**Review by joris-burgers — COMMENTED** [review]
Nice work, the code looks like what it supposed to do! I have some minor style points, but nothing major.

File: src/Logic.hs
Comment (FPtje) [inline src/Logic.hs:602]: Nit: Double space

File: src/Project.hs
Comment (FPtje) [inline src/Project.hs:335]: Nit: I think one space before the `->` would be best here

File: tests/Spec.hs
Comment (FPtje) [inline tests/Spec.hs:746]: These are good tests!

**Review by FPtje — CHANGES_REQUESTED** [review]
Nice work! Some more comments!

File: src/Logic.hs
Comment (adrianotm) [reply src/Logic.hs:264]: It would leave a comment if the PR is the integration candidate. In that case it will also clear the integration candidate. 

The "Abandoning this pull request because it was closed." message is not appropriate though. I think fixing issue #66 should display an appropriate message for this case as well.

File: tests/Spec.hs
Comment (adrianotm) [reply tests/Spec.hs:746]: Thanks! :smile:

File: src/Logic.hs
Comment (FPtje) [reply src/Logic.hs:264]: This change does introduce the same bug in a new location, though.

By substituting the closed/opened function definitions, you can leave out the parts that are wrong:

```haskell
    Pr.insertPullRequest prId branch baseBranch sha title author . Pr.deletePullRequest prId <$>
      case Pr.integrationCandidate state of
        Just candidatePr | candidatePr == prId ->
          pure state { Pr.integrationCandidate = Nothing }
        _notCandidatePr -> pure state
```

That should then also fix #66, right?

Comment (adrianotm) [reply src/Logic.hs:264]: Yes, this should fix it. But I think it's useful if hoff shows a message that it stopped the integration. We can use different messages for different cases.

Comment (FPtje) [reply src/Logic.hs:264]: Oh wait, no, that would remove all the comments. In that case, you can just change the comment to `Abandoning pull request because it changed after approval`. That would cover both having commits and the base branch changed :+1:

Comment (adrianotm) [reply src/Logic.hs:264]: But if we changed the comment completely, we would get an incorrect message when a user actually closes a PR. I think maybe we should change the message depending on the situation. I guess it would be useful to know what caused the abandoning.

Comment (adrianotm) [reply src/Logic.hs:264]: Oh, I guess you are talking about inlining the `handlePullRequestClosed` in `clearPullRequest`. I guess that's going to work. I still think showing different messages would be nice, but idk...

Comment (FPtje) [reply src/Logic.hs:264]: Yeah, indeed. For closed pull requests it would still call `handlePullRequestClosed` with the correct message. In here the message is indeed generic over rebase and commits pushed, but I think that's fine. The change can be seen in the PR right above the comment that would be posted.

Comment (adrianotm) [reply src/Logic.hs:264]: That's true, good point.

File: src/Git.hs
Comment (joris-burgers) [inline src/Git.hs:78]: Thanks for also fixing the existing comments!

File: src/Logic.hs
Comment (joris-burgers) [inline src/Logic.hs:214]: Another Haddock convention, if you change the comment style to `-- ^`, it attaches the comment to the constructor that came before, which is really useful if you hover over a constructor in your editor  to see what everything means.

File: src/Logic.hs
Comment (joris-burgers) [reply src/Logic.hs:254]: Extra nit: You can use a `-- |` here as well.

File: src/Project.hs
Comment (joris-burgers) [inline src/Project.hs:94]: nit: Please use `-- ^` here as well.

File: src/Project.hs
Comment (joris-burgers) [inline src/Project.hs:77]: nit: Use `-- |` here please.

File: tests/Spec.hs
Comment (joris-burgers) [inline tests/Spec.hs:629]: nit: You can use `masterBranch` here as well.

**Review by joris-burgers — COMMENTED** [review]
I left some more (very minor) style comments, but overall the code looks good.

File: src/Logic.hs
Comment (adrianotm) [reply src/Logic.hs:214]: Thanks for the tip, but I think I'd need a new PR to fix all the comments. :sweat_smile:

File: src/Logic.hs
Comment (adrianotm) [reply src/Logic.hs:264]: I suggest implementing this in another PR. I'll need to add a test for when a commit is made as well and I am afraid this PR will grow outside of its scope. Wdyt?

Comment (FPtje) [reply src/Logic.hs:264]: Alright! Good call on limiting the scope :+1:

**Review by joris-burgers — APPROVED** [review]
Thanks for fixing, LGTM from my side.

File: tests/Spec.hs
Comment (marcin-rzeznicki) [inline tests/Spec.hs:741]: Thx for adding

**Review by marcin-rzeznicki — APPROVED** [review]
I am content. Thx for adding good tests

**Review by FPtje — APPROVED** [review]
Nice work! LGTM!

**Comment by adrianotm** [comment]
@OpsBotPrime merge

**Comment by OpsBotPrime** [comment]
Pull request approved for merge by @adrianotm, rebasing now.

**Comment by OpsBotPrime** [comment]
Rebased as 20c6e6708bf43219bab560aedd32ed3009f7dca0, waiting for CI …

**Pull request merged**
